libmamba/include/mamba/core/channel.hpp

```diff
@@ -55,40 +55,55 @@ namespace mamba
                     -> std::optional<std::string_view>;
             };
 
+            template <typename Key, typename Value>
+            using name_map = util::weakening_map<std::unordered_map<Key, Value>, NameWeakener>;
+
             using platform_list = util::flat_set<std::string>;
             using channel_list = std::vector<Channel>;
-            using channel_map = util::weakening_map<std::unordered_map<std::string, Channel>, NameWeakener>;
-            using multichannel_map = std::unordered_map<std::string, channel_list>;
+            using channel_map = name_map<std::string, Channel>;
+            using multichannel_map = name_map<std::string, channel_list>;
 
             const platform_list& platforms;
             const specs::CondaURL& channel_alias;
             const channel_map& custom_channels;
+            const multichannel_map& custom_multichannels;
             const specs::AuthenticationDataBase& auth_db;
 
             // TODO add CWD and home
         };
 
-        using platform_list = util::flat_set<std::string>;
+        using platform_list = ResolveParams::platform_list;
+        using channel_list = ResolveParams::channel_list;
 
-        [[nodiscard]] static auto resolve(specs::ChannelSpec spec, ResolveParams params) -> Channel;
+        [[nodiscard]] static auto resolve(specs::ChannelSpec spec, ResolveParams params)
+            -> channel_list;
 
         Channel(specs::CondaURL url, std::string display_name, util::flat_set<std::string> platforms = {});
 
         [[nodiscard]] auto url() const -> const specs::CondaURL&;
+        auto clear_url() -> const specs::CondaURL;
         void set_url(specs::CondaURL url);
 
         [[nodiscard]] auto platforms() const -> const platform_list&;
+        auto clear_platforms() -> platform_list;
         void set_platforms(platform_list platforms);
 
         [[nodiscard]] auto display_name() const -> const std::string&;
+        auto clear_display_name() -> std::string;
         void set_display_name(std::string display_name);
 
-        std::string base_url() const;
-        std::string platform_url(std::string_view platform, bool with_credential = true) const;
+        [[nodiscard]] auto url_equivalent_with(const Channel& other) const -> bool;
+
+        [[nodiscard]] auto is_equivalent_to(const Channel& other) const -> bool;
+
+        [[nodiscard]] auto contains_equivalent(const Channel& other) const -> bool;
+
+        [[nodiscard]] auto
+        platform_url(std::string_view platform, bool with_credential = true) const -> std::string;
         // The pairs consist of (platform,url)
-        util::flat_set<std::pair<std::string, std::string>>
-        platform_urls(bool with_credential = true) const;
-        util::flat_set<std::string> urls(bool with_credential = true) const;
+        [[nodiscard]] auto platform_urls(bool with_credential = true) const
+            -> util::flat_set<std::pair<std::string, std::string>>;
```

Comment:
The return type deserves a typename

Reply:
This part still needs improvement overall, such as not converting URL to strings.
However, I will first rework the tests, since it is heavily (and unnecessarily anymore) used there.